### PR TITLE
add component certification doc

### LINKED ISSUE #935: Document components maturity status
Components can have the following maturity statuses:

* `alpha`
* `beta`
* `stable`

This issue is to update each component with its maturity level, and provide a high-level explanation about what the different statuses mean.

### PR DESCRIPTION
**Pull request by mukundansundar**
Thank you for helping make the Dapr documentation better!

**Please follow this checklist before submitting:**

- [x] [Read the contribution guide](https://docs.dapr.io/contributing/contributing-docs/)
- [x] Commands include options for Linux, MacOS, and Windows within codetabs
- [x] New file and folder names are globally unique
- [x] Page references use shortcodes instead of markdown or URL links
- [x] Images use HTML style and have alternative text
- [x] Places where multiple code/command options are given have codetabs

In addition, please fill out the following to help reviewers understand this pull request:

## Description

Added the component certification doc

## Issue reference

closes #935

### CODE REVIEW

File: daprdocs/content/en/operations/components/certification-lifecycle.md [hunk 1]
```diff
@@ -0,0 +1,94 @@
...
+- Not recommended for use in production
+
+All components generally start at the Alpha stage.
+
+### Beta
+
+- Component must pass all the component conformance tests defined to satisfy the component specification
+- The component conformance tests have been run in a Dapr maintainers managed environment
+- Component contains a record of the conformance test result reviewed and approved by Dapr maintainers with specific components-contrib version
+- Not recommended for use in production
```
Comment (yaron2) [inline daprdocs/content/en/operations/components/certification-lifecycle.md]: I would reword this to: `Recommended for only non-business-critical uses because of potential for incompatible changes in subsequent releases.`. The text is taken from kubernetes.io, I would rephrase a little.

File: daprdocs/content/en/operations/components/certification-lifecycle.md [hunk 2]
```diff
@@ -0,0 +1,94 @@
...
+type: docs
+title: "Certification lifecycle"
+linkTitle: "Certification lifecycle"
+weight: 200
+description: "The certification lifecycle of the different components available in Dapr"
+---
+
+## Overview
+
+Dapr uses a modular design where functionality is delivered as a component. Each component has an interface definition.  All of the components are pluggable so that in ideal scenarios, you can swap out one component with the same interface for another. Each component that is used in production, needs to matain a certain set of technical requirements that ensure the functional compatibility and robustness of the component.
```
Comment (artursouza) [inline daprdocs/content/en/operations/components/certification-lifecycle.md]: Typo in `matain`

File: daprdocs/content/en/operations/components/certification-lifecycle.md [hunk 3]
```diff
@@ -0,0 +1,94 @@
...
+
+In general a component needs to be: 
+- compliant with the defined Dapr interfaces 
+- functionally correct and robust
+- well documented and maintained
+
+To make sure a component conforms to the standards set by Dapr, there are a set of tests run against a component in a Dapr maintainers managed environment. Once the tests pass consistently for a certain number of runs, the maturity level can be determined for a component.
```
Comment (artursouza) [inline daprdocs/content/en/operations/components/certification-lifecycle.md]: Please, be precise in "certain number of runs" - there should be a number here. Or remove this requirement.

File: daprdocs/content/en/operations/components/certification-lifecycle.md [hunk 4]
```diff
@@ -0,0 +1,94 @@
...
+
+### General Availability (GA)
+
+- Has at least two different users in production
+- A GA component has a defined maintainer in the Dapr community or the Dapr maintainers
+- Component is well documented, tested and maintained across multiple versions of components-contrib repo
+
+## Conformance tests 
+
+Each component in the [components-contrib](https://github.com/dapr/components-contrib) repository needs to adhere to a set of interfaces defined by Dapr. Conformance tests are tests that are run on these component definitions with their associated backing services such that the component is tested to be conformant with the Dapr interface specifications.
```
Comment (artursouza) [inline daprdocs/content/en/operations/components/certification-lifecycle.md]: mention behavior too. There is some behavior that is validated in the conf tests.

File: daprdocs/content/en/operations/components/certification-lifecycle.md [hunk 5]
```diff
@@ -0,0 +1,94 @@
...
+- The tests should validate the functional behavior and robustness of component based on the component specification
+- All the details needed to reproduce the tests will be added as part of the component conformance test documentation
+
+## Component certification process
+
+### New component certification
+
+For a new component a request for component certification follows these steps:
+- This applies for either certifying a component as Beta from Alpha or GA from Beta stages
+- An issue is created with a request for certification of the component with the current and the new certification levels
```
Comment (artursouza) [inline daprdocs/content/en/operations/components/certification-lifecycle.md:76]: Create a template in GitHub please.

File: daprdocs/content/en/operations/components/certification-lifecycle.md [hunk 6]
```diff
@@ -0,0 +1,94 @@
...
+- The component owner/maintainer submits a PR for integrating the component to run with the defined conformance test suite
```
Comment (artursouza) [inline daprdocs/content/en/operations/components/certification-lifecycle.md]: How do we know who "owns" a component? Is there a owners file somewhere?

File: daprdocs/content/en/operations/components/certification-lifecycle.md [hunk 7]
```diff
@@ -0,0 +1,94 @@
...
+- The component owner/maintainer details the environment setup in the issue created, so that a Dapr maintainer can setup the service in a managed environment
+- After the environment setup is complete, Dapr matainers review the PR and if approved merges that PR
+- Dapr maintainers review functional correctness with the test being run in a Dapr team maintained environment
+- Dapr maintainers update component status document categorized by Dapr Runtime version, this will be done as part of the release process in the following release of Dapr runtime
+
+### Existing GA certified component
+
+For an existing GA certified component:
+- If there only internal method (not interface methods) signature changes but the behavior has not changed eg: adding logging, then there is no need to re-certify the component
+- The component need to be recertified in these scenarios:
```
Comment (artursouza) [inline daprdocs/content/en/operations/components/certification-lifecycle.md]: bug fix? new feature?

File: daprdocs/content/en/operations/components/certification-lifecycle.md [hunk 8]
```diff
@@ -0,0 +1,94 @@
...
+  - if there are dependency changes, eg: version of redis client used
+  - if the component needs to be certified with a new version of backing service
+- For recertification the tests are run for the component with the new versions of the client and/or the backing service and results verified
+- The new component certification process is followed in these scenarios:
```
Comment (artursouza) [inline daprdocs/content/en/operations/components/certification-lifecycle.md]: What is the difference from this condition and the condition 2 steps above?

File: daprdocs/content/en/operations/components/certification-lifecycle.md [hunk 9]
```diff
@@ -0,0 +1,94 @@
...
+  - if the behavior of the component is changing
+  - if there are breaking changes in terms of function signature/interface changes
+
+For a component to be certified tests will be run in a Dapr team maintained environment.
```
Comment (artursouza) [inline daprdocs/content/en/operations/components/certification-lifecycle.md]: ...in an environment maintained by the Dapr team.

File: daprdocs/content/en/operations/components/setup-pubsub/supported-pubsub/_index.md [hunk 10]
```diff
@@ -1,39 +1,47 @@
...
-| [NATS Streaming]({{< ref setup-nats-streaming.md >}}) | Alpha  | 
-| [Pulsar]({{< ref setup-pulsar.md >}})                 | Alpha  | 
-| [RabbitMQ]({{< ref setup-rabbitmq.md >}})             | Alpha  | 
-| [Redis Streams]({{< ref setup-redis-pubsub.md >}})    | Alpha  | 
+| Name                                                  | Status | Component version | Since |
+|-------------------------------------------------------|--------| -----| ------------- |
+| [Apache Kafka]({{< ref setup-apache-kafka.md >}})     | Alpha  | v1 | 1.0 |
+| [Hazelcast]({{< ref setup-hazelcast.md >}})           | Alpha  | v1 | 1.0 |
+| [MQTT]({{< ref setup-mqtt.md >}})                     | Alpha  | v1 | 1.0 |
+| [NATS Streaming]({{< ref setup-nats-streaming.md >}}) | Alpha  | v1 | 1.0 |
```
Comment (artursouza) [inline daprdocs/content/en/operations/components/setup-pubsub/supported-pubsub/_index.md]: It passes conf tests, does not mean Beta?

File: daprdocs/content/en/operations/components/setup-pubsub/supported-pubsub/_index.md [hunk 11]
```diff
@@ -1,39 +1,47 @@
...
-| [Apache Kafka]({{< ref setup-apache-kafka.md >}})     | Alpha  | 
-| [Hazelcast]({{< ref setup-hazelcast.md >}})           | Alpha  | 
-| [MQTT]({{< ref setup-mqtt.md >}})                     | Alpha  | 
-| [NATS Streaming]({{< ref setup-nats-streaming.md >}}) | Alpha  | 
-| [Pulsar]({{< ref setup-pulsar.md >}})                 | Alpha  | 
-| [RabbitMQ]({{< ref setup-rabbitmq.md >}})             | Alpha  | 
-| [Redis Streams]({{< ref setup-redis-pubsub.md >}})    | Alpha  | 
+| Name                                                  | Status | Component version | Since |
+|-------------------------------------------------------|--------| -----| ------------- |
+| [Apache Kafka]({{< ref setup-apache-kafka.md >}})     | Alpha  | v1 | 1.0 |
```
Comment (artursouza) [inline daprdocs/content/en/operations/components/setup-pubsub/supported-pubsub/_index.md]: Beta?

File: daprdocs/content/en/operations/components/setup-secret-store/supported-secret-stores/_index.md [hunk 12]
```diff
@@ -7,30 +7,39 @@ description: The supported secret stores that interface with Dapr
...
-| Name                                                              | Status                       |
-|-------------------------------------------------------------------|------------------------------|
-| [Local environment variables]({{< ref envvar-secret-store.md >}}) | GA (For local development)   |
-| [Local file]({{< ref file-secret-store.md >}})                    | GA (For local development)   |
-| [HashiCorp Vault]({{< ref hashicorp-vault.md >}})                 | Alpha                        |
-| [Kubernetes secrets]({{< ref kubernetes-secret-store.md >}})      | Alpha                        |
+| Name                                                              | Status                       | Component version | Since |
+|-------------------------------------------------------------------|------------------------------| ---------------- |-- |
+| [Local environment variables]({{< ref envvar-secret-store.md >}}) | GA (For local development)   | v1 | 1.0 |
+| [Local file]({{< ref file-secret-store.md >}})                    | GA (For local development)   | v1 | 1.0 |
```
Comment (artursouza) [inline daprdocs/content/en/operations/components/setup-secret-store/supported-secret-stores/_index.md]: This is odd. GA for local development? Why not just call it Beta forever.

**Review by artursouza — CHANGES_REQUESTED** [review]

File: daprdocs/content/en/operations/components/certification-lifecycle.md (hunk 6 above)
Comment (msfussell) [reply daprdocs/content/en/operations/components/certification-lifecycle.md]: This was my question. Is is worth having tracking issues (which are closed by can be updated) for each component? Or is this best in docs? The advantage of the issue, is that this maintains the history.

File: daprdocs/content/en/operations/components/setup-bindings/supported-bindings/_index.md [hunk 13]
```diff
@@ -3,30 +3,43 @@ type: docs
...
-| [Postmark]({{< ref postmark.md >}})       |    | ✅ | Alpha |
-| [RabbitMQ]({{< ref rabbitmq.md >}})   | ✅ | ✅ | Alpha |
-| [Redis]({{< ref redis.md >}})         |    | ✅ | Alpha |
-| [SMTP]({{< ref smtp.md >}})         |    | ✅ | Alpha |
-| [Twilio]({{< ref twilio.md >}})       |    | ✅ | Alpha |
-| [Twitter]({{< ref twitter.md >}})       | ✅ | ✅ | Alpha |
-| [SendGrid]({{< ref sendgrid.md >}})       |    | ✅ | Alpha |
+| Name | Input<br>Binding | Output<br>Binding | Status | Component version | Since |
+|------|:----------------:|:-----------------:|--------|-------- | ---------|
+| [Apple Push Notifications (APN)]({{< ref apns.md >}}) |  | ✅ | Alpha | v1 | 1.0 |
```
Comment (msfussell) [inline daprdocs/content/en/operations/components/setup-bindings/supported-bindings/_index.md:26]: Its odd that the Component and Runtime Since version are in different formats. I would change all the components from v1 to be 1.0 so that this is the same as the runtime. Or have v1.0 for all of them (but the v is redundant here)

File: daprdocs/content/en/operations/components/setup-pubsub/supported-pubsub/_index.md [hunk 14]
```diff
@@ -1,39 +1,47 @@
...
-| Name                                                  | Status |
-|-------------------------------------------------------|--------|
-| [Apache Kafka]({{< ref setup-apache-kafka.md >}})     | Alpha  | 
-| [Hazelcast]({{< ref setup-hazelcast.md >}})           | Alpha  | 
-| [MQTT]({{< ref setup-mqtt.md >}})                     | Alpha  | 
-| [NATS Streaming]({{< ref setup-nats-streaming.md >}}) | Alpha  | 
-| [Pulsar]({{< ref setup-pulsar.md >}})                 | Alpha  | 
-| [RabbitMQ]({{< ref setup-rabbitmq.md >}})             | Alpha  | 
-| [Redis Streams]({{< ref setup-redis-pubsub.md >}})    | Alpha  | 
+| Name                                                  | Status | Component version | Since |
```
Comment (msfussell) [inline daprdocs/content/en/operations/components/setup-pubsub/supported-pubsub/_index.md:21]: Same comment here on version format above.

File: daprdocs/content/en/operations/components/setup-secret-store/supported-secret-stores/_index.md [hunk 15]
```diff
@@ -7,30 +7,39 @@ description: The supported secret stores that interface with Dapr
...
 
-| Name                                                              | Status                       |
-|-------------------------------------------------------------------|------------------------------|
-| [Local environment variables]({{< ref envvar-secret-store.md >}}) | GA (For local development)   |
-| [Local file]({{< ref file-secret-store.md >}})                    | GA (For local development)   |
-| [HashiCorp Vault]({{< ref hashicorp-vault.md >}})                 | Alpha                        |
-| [Kubernetes secrets]({{< ref kubernetes-secret-store.md >}})      | Alpha                        |
+| Name                                                              | Status                       | Component version | Since |
+|-------------------------------------------------------------------|------------------------------| ---------------- |-- |
+| [Local environment variables]({{< ref envvar-secret-store.md >}}) | GA (For local development)   | v1 | 1.0 |
```
Comment (msfussell) [inline daprdocs/content/en/operations/components/setup-secret-store/supported-secret-stores/_index.md]: Same comment here on version format above.

File: daprdocs/content/en/operations/components/setup-state-store/supported-state-stores/_index.md [hunk 16]
```diff
@@ -1,40 +1,52 @@
...
-| [Hazelcast]({{< ref setup-hazelcast.md >}})                    | ✅   | ❌            |  Alpha  |
-| [Memcached]({{< ref setup-memcached.md >}})                    | ✅   | ❌            |  Alpha  |
-| [MongoDB]({{< ref setup-mongodb.md >}})                        | ✅   | ✅            |  Alpha  |
-| [MySQL]({{< ref setup-mysql.md >}})                            | ✅   | ✅            |  Alpha  |
-| [PostgreSQL]({{< ref setup-postgresql.md >}})                  | ✅   | ✅            |  Alpha  |
-| [Redis]({{< ref setup-redis.md >}})                            | ✅   | ✅            |  Alpha  |
-| [Zookeeper]({{< ref setup-zookeeper.md >}})                    | ✅   | ❌            |  Alpha  |
+| Name                                                           | CRUD | Transactional <br/> (needed for Actors) | Status | Component version | Since |
+|----------------------------------------------------------------|------|---------------|--------| ---- |--------------|
+| [Aerospike]({{< ref setup-aerospike.md >}})                    | ✅   | ❌            |  Alpha  | v1 | 1.0 |
```
Comment (msfussell) [inline daprdocs/content/en/operations/components/setup-state-store/supported-state-stores/_index.md]: Same comment here on version format above.

File: daprdocs/content/en/operations/components/setup-bindings/supported-bindings/_index.md (hunk 13 above)
Comment (artursouza) [reply daprdocs/content/en/operations/components/setup-bindings/supported-bindings/_index.md:26]: These are versioned independently, actually. And there is no v1.1 component, for example. It is just a coincidence right now because we are starting it.

**Comment by msfussell** [comment]
@mukundansundar - Any updates here?

File: daprdocs/content/en/operations/components/certification-lifecycle.md (hunk 5 above)
Comment (mukundansundar) [reply daprdocs/content/en/operations/components/certification-lifecycle.md:76]: will do

File: daprdocs/content/en/operations/components/certification-lifecycle.md (hunk 6 above)
Comment (mukundansundar) [reply daprdocs/content/en/operations/components/certification-lifecycle.md]: I think having an issue tracking who own what component will be beneficial ... For tracking and maintaining history.
Or one more idea, we could have in each component after the license a comment line containing the maintainers of a component. It lives with the code and would evolve with the code then.

File: daprdocs/content/en/operations/components/certification-lifecycle.md (hunk 8 above)
Comment (mukundansundar) [reply daprdocs/content/en/operations/components/certification-lifecycle.md]: Here the behavior of the component is changing and there might be breaking changes associated with that. In essence a version update of the component is there, then treat it as a `new component` certification.

File: daprdocs/content/en/operations/components/setup-bindings/supported-bindings/_index.md (hunk 13 above)
Comment (mukundansundar) [reply daprdocs/content/en/operations/components/setup-bindings/supported-bindings/_index.md:26]: @msfussell There is a reason why the format is different, in the component YAML, for the version field, a component has the version as `v0`,  `v1` etc. I wanted to maintain that format here also.

File: daprdocs/content/en/operations/components/setup-pubsub/supported-pubsub/_index.md (hunk 14 above)
Comment (mukundansundar) [reply daprdocs/content/en/operations/components/setup-pubsub/supported-pubsub/_index.md:21]: see comment above

File: daprdocs/content/en/operations/components/setup-secret-store/supported-secret-stores/_index.md (hunk 15 above)
Comment (mukundansundar) [reply daprdocs/content/en/operations/components/setup-secret-store/supported-secret-stores/_index.md]: see comment above.

File: daprdocs/content/en/operations/components/setup-state-store/supported-state-stores/_index.md (hunk 16 above)
Comment (mukundansundar) [reply daprdocs/content/en/operations/components/setup-state-store/supported-state-stores/_index.md]: see comment above

**Review by yaron2 — APPROVED** [review]

File: daprdocs/content/en/operations/components/certification-lifecycle.md (hunk 6 above)
Comment (artursouza) [reply daprdocs/content/en/operations/components/certification-lifecycle.md]: What about an owners.txt file in the component's directory?

File: daprdocs/content/en/operations/components/certification-lifecycle.md (hunk 8 above)
Comment (artursouza) [reply daprdocs/content/en/operations/components/certification-lifecycle.md]: Can you reword it to make it clear? It was not that obvious.

File: daprdocs/content/en/operations/components/certification-lifecycle.md (hunk 6 above)
Comment (artursouza) [reply daprdocs/content/en/operations/components/certification-lifecycle.md]: Or `maintainers.txt` since `owners` might not be good terminology in open source.

File: daprdocs/content/en/operations/components/certification-lifecycle.md (hunk 8 above)
Comment (mukundansundar) [reply daprdocs/content/en/operations/components/certification-lifecycle.md]: can you read it again. I have added the version change part.

File: daprdocs/content/en/operations/components/setup-bindings/supported-bindings/_index.md (hunk 13 above)
Comment (msfussell) [reply daprdocs/content/en/operations/components/setup-bindings/supported-bindings/_index.md:26]: Yes. I got this wrong. This is correct

**Review by msfussell — APPROVED** [review]

File: daprdocs/content/en/operations/components/certification-lifecycle.md [hunk 17]
```diff
@@ -0,0 +1,95 @@
...
+### Test requirements
+
+- The tests should validate the functional behavior and robustness of component based on the component specification
+- All the details needed to reproduce the tests are added as part of the component conformance test documentation
+
+## Component certification process
+
+### New component certification: Alpha->Beta or Beta->GA
+
+For a new component a request for component certification follows these steps. This applies for either certifying a component going from Beta from Alpha or GA from Beta.
```
Comment (artursouza) [inline daprdocs/content/en/operations/components/certification-lifecycle.md]: For a new component, a request for component ...

Comment (artursouza) [reply daprdocs/content/en/operations/components/certification-lifecycle.md]: Can you rephrase this? "This applies for either certifying a component going from Beta from Alpha or GA from Beta."

File: daprdocs/content/en/operations/components/certification-lifecycle.md [hunk 18]
```diff
@@ -0,0 +1,95 @@
...
+- An issue is created with a request for certification of the component with the current and the new certification levels
+- A user of a component user submits a PR for integrating the component to run with the defined conformance test suite
```
Comment (artursouza) [inline daprdocs/content/en/operations/components/certification-lifecycle.md]: "A user of a component user " ? Can you rephrase this?

File: daprdocs/content/en/operations/components/certification-lifecycle.md [hunk 19]
```diff
@@ -0,0 +1,95 @@
...
+- The user details the environment setup in the issue created, so that a Dapr maintainer can setup the service in a managed environment
+- After the environment setup is complete, Dapr maintainers review the PR and if approved merges that PR
+- Dapr maintainers review functional correctness with the test being run in a Dapr team maintained environment
```
Comment (artursouza) [inline daprdocs/content/en/operations/components/certification-lifecycle.md]: Dapr team's maintained environment.

File: daprdocs/content/en/operations/components/certification-lifecycle.md [hunk 20]
```diff
@@ -0,0 +1,95 @@
...
+- Dapr maintainers update the component status document categorized by Dapr Runtime version. This is done as part of the release process in the next release of Dapr runtime
+
+### Existing GA certified component
+
+For an existing GA certified component:
+- If there only internal method (not interface methods) signature changes but the behavior has not changed (for example adding new logging), then there is no need to re-certify the component
```
Comment (artursouza) [inline daprdocs/content/en/operations/components/certification-lifecycle.md]: This contradicts the section below where "...component has bug fixes or feature improvements". How can a change not be a fix or an improvement?

**Review by artursouza — CHANGES_REQUESTED** [review]

**Review by artursouza — APPROVED** [review]
